### PR TITLE
CNS-925-slashing-gov-proposal-testnet

### PR DESCRIPTION
**Pull request by Yaroms**

### CODE REVIEW

File: x/pairing/proposal_handler.go [hunk 1]
```diff
@@ -66,6 +66,10 @@ func handleUnstakeProposal(ctx sdk.Context, k keeper.Keeper, p *types.UnstakePro
 		}
 	}
 
+	for _, delegatorSlashing := range p.DelegatorsSlashing {
+		err := k.SlashDelegator(ctx, delegatorSlashing)
+		_ = err
```
Comment (oren-lava) [inline x/pairing/proposal_handler.go:71]: return the error

File: x/pairing/keeper/unstaking.go [hunk 2]
```diff
@@ -142,3 +143,81 @@ func (k Keeper) UnstakeEntryForce(ctx sdk.Context, chainID, provider, unstakeDes
...
+		if !found {
+			continue
+		}
+		amount := validator.TokensFromShares(delegation.Shares).TruncateInt()
+		if totalUnbonding.LT(amount) {
+			amount = totalUnbonding
+		}
+		shares, err := validator.SharesFromTokensTruncated(amount)
+		if err != nil {
+			return utils.LavaFormatWarning("blablabla", err,
```
Comment (oren-lava) [inline x/pairing/keeper/unstaking.go:195]: blabla

File: x/pairing/keeper/unstaking.go [hunk 3]
```diff
@@ -142,3 +143,81 @@ func (k Keeper) UnstakeEntryForce(ctx sdk.Context, chainID, provider, unstakeDes
...
+				utils.Attribute{Key: "address", Value: delAddr},
+			)
+		}
+
+		totalUnbonding = total.Sub(totalUnbonding)
```
Comment (oren-lava) [inline x/pairing/keeper/unstaking.go:200]: totalUnbonding = totalUnbonding.Sub(amount)

File: x/pairing/keeper/unstaking.go [hunk 4]
```diff
@@ -142,3 +143,81 @@ func (k Keeper) UnstakeEntryForce(ctx sdk.Context, chainID, provider, unstakeDes
...
+	// we need to unbond from the delegator by force the amount left
+	delegations := k.stakingKeeper.GetAllDelegatorDelegations(ctx, delAddr)
+	totalUnbonding := total
+	for _, delegation := range delegations {
+		validator, found := k.stakingKeeper.GetValidator(ctx, delegation.GetValidatorAddr())
+		if !found {
+			continue
+		}
+		amount := validator.TokensFromShares(delegation.Shares).TruncateInt()
+		if totalUnbonding.LT(amount) {
```
Comment (oren-lava) [inline x/pairing/keeper/unstaking.go:190]: use math.min

File: x/pairing/keeper/unstaking.go [hunk 5]
```diff
@@ -142,3 +143,81 @@ func (k Keeper) UnstakeEntryForce(ctx sdk.Context, chainID, provider, unstakeDes
...
+			)
+		}
+
+		if totalUnbonding.IsZero() {
+			break
+		}
+	}
+
+	if !totalUnbonding.IsZero() {
+		return utils.LavaFormatWarning("delegator does not have enough delegations to slash", err,
```
Comment (oren-lava) [inline x/pairing/keeper/unstaking.go:214]: instead just print warning that you slash a different amount

**Review by oren-lava — CHANGES_REQUESTED** [review]

File: x/pairing/proposal_handler.go (hunk 1 above)
Comment (Yaroms) [reply x/pairing/proposal_handler.go:71]: done

**Review by oren-lava — APPROVED** [review]

File: proto/lavanet/lava/pairing/unstake_proposal.proto [hunk 6]
```diff
@@ -5,16 +5,24 @@ option go_package = "github.com/lavanet/lava/x/pairing/types";
...
 }
 
 message ProviderUnstakeInfo {
   string provider = 1;
   string chain_id = 2;
 }
+
+message DelegatorSlashing {
+  string delegator = 1;
+  cosmos.base.v1beta1.Coin delegate_limit = 2 [(gogoproto.nullable) = false];
```
Comment (Yaroms) [inline proto/lavanet/lava/pairing/unstake_proposal.proto:27]: change name

File: x/pairing/keeper/unstaking.go [hunk 7]
```diff
@@ -142,3 +143,80 @@ func (k Keeper) UnstakeEntryForce(ctx sdk.Context, chainID, provider, unstakeDes
...
+
+			if total.IsZero() {
+				return
+			}
+		}
+	}
+
+	// try slashing existing unbondings
+	slashUnbonding()
+	if total.IsZero() {
```
Comment (omerlavanet) [inline x/pairing/keeper/unstaking.go:178]: less eq

File: x/pairing/keeper/unstaking.go (hunk 5 above)
Comment (omerlavanet) [reply x/pairing/keeper/unstaking.go:214]: agreed here

**Review by omerlavanet — APPROVED** [review]

**Pull request merged**